dape.el

```diff
@@ -916,199 +895,140 @@ On SKIP-PROCESS-BUFFERS skip deletion of buffers which has processes."
+    (when (stringp seq) ;; dirty dirty netcoredbg
+      (setq seq (string-to-number seq)))
+    (cond ((string= type "event")
+           `(:method ,event :params ,body))
+          ((string= type "response")
+           `(:id ,request_seq :result ,dap-message))
+          (command
+           `(:id ,seq :method ,command :params ,arguments)))))
 
 
 ;;; Outgoing requests
 
-(defconst dape--timeout 5
-  "Time before dape starts to complain about missing responses.")
-
-(defun dape--create-timer (process seq)
-  "Create SEQ request timeout timer for PROCESS."
-  (puthash seq
-           (run-with-timer dape--timeout
-                           nil
-                           (dape--callback
-                            (dape--debug 'error
-                                         "Timeout for reached for seq %d"
-                                         seq)
-                            (when (dape--live-process t)
-                              (dape--update-state 'timed-out))
-                            (remhash seq dape--timers)
-                            (when-let ((cb (gethash seq dape--cb)))
-                              (remhash seq dape--cb)
-                              (funcall cb process nil nil nil)))
-                           process)
-           dape--timers))
-
-(defun dape-send-object (process &optional seq object)
-  "Helper for `dape-request' to send SEQ request with OBJECT to PROCESS."
-  (let* ((object (if seq (plist-put object :seq seq) object))
-         (json (json-serialize object :false-object nil))
-         (string (format "Content-Length: %d\r\n\r\n%s" (length json) json)))
-    (dape--debug 'io "Sending:\n%S" object)
-    (condition-case err
-        (process-send-string process string)
-      (error (dape--debug 'error "%s"
-                          (error-message-string err))))))
-
-(defun dape-request (process command arguments &optional cb skip-timeout)
-  "Send request COMMAND to PROCESS with ARGUMENTS.
-If CB set, invoke CB on response.
-If SKIP-TIMEOUT non nil skip timeout handler creation.
-See `dape--callback' for expected function signature."
-  (let ((seq (setq dape--seq (1+ dape--seq)))
-        (object (and arguments (list :arguments arguments))))
-    (unless skip-timeout
-      (dape--create-timer process seq))
-    (when cb
-      (puthash seq cb dape--cb))
-    (dape-send-object process
-                      seq
-                      (thread-first object
-                                    (plist-put :type "request")
-                                    (plist-put :command command)))))
-
-(defun dape--initialize (process)
-  "Initialize and launch/attach session for PROCESS."
-  (dape--with dape-request (process
+(defun dape-request (conn command arguments &optional cb)
+  "Send request with COMMAND and ARGUMENTS to adapter CONN.
+If callback function CB is supplied, it's called on timeout
+and success.  See `dape--callback' for signature."
+  (jsonrpc-async-request conn command arguments
```

Comment:
I would consider using `jsonrpc-request` as my main workhorse.  `jsonrpc-async-request` only seems to be necessary in the REPL, as far as I can tell, because you don't want to block the UI but you also don't want to cancel the request either.